{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['mesh',\n",
       " 'go',\n",
       " '.DS_Store',\n",
       " 'id_mapper',\n",
       " 'annotation',\n",
       " 'kegg',\n",
       " 'interpro',\n",
       " 'reactome',\n",
       " 'msigdb']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.listdir('data/tmp/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### create species table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name_short</th>\n",
       "      <th>ek_species</th>\n",
       "      <th>name_long</th>\n",
       "      <th>current_gtf</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>bta</td>\n",
       "      <td>0</td>\n",
       "      <td>Bos_taurus</td>\n",
       "      <td>ARS-UCD1.2.107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cap</td>\n",
       "      <td>1</td>\n",
       "      <td>Capra_hircus</td>\n",
       "      <td>ARS1.107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ovi</td>\n",
       "      <td>2</td>\n",
       "      <td>Ovis_aries</td>\n",
       "      <td>Oar_v3.1.107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gal</td>\n",
       "      <td>3</td>\n",
       "      <td>Gallus_gallus</td>\n",
       "      <td>gca000002315v5.GRCg6a.107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sus</td>\n",
       "      <td>4</td>\n",
       "      <td>Sus_scrofa</td>\n",
       "      <td>Sscrofa11.1.107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>equ</td>\n",
       "      <td>5</td>\n",
       "      <td>Equus_caballus</td>\n",
       "      <td>EquCab3.0.107</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  name_short  ek_species       name_long                current_gtf\n",
       "0        bta           0      Bos_taurus             ARS-UCD1.2.107\n",
       "1        cap           1    Capra_hircus                   ARS1.107\n",
       "2        ovi           2      Ovis_aries               Oar_v3.1.107\n",
       "3        gal           3   Gallus_gallus  gca000002315v5.GRCg6a.107\n",
       "4        sus           4      Sus_scrofa            Sscrofa11.1.107\n",
       "5        equ           5  Equus_caballus              EquCab3.0.107"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_species = {\n",
    "        'bta':[0,'Bos_taurus','ARS-UCD1.2.107'],\n",
    "        'cap':[1,'Capra_hircus','ARS1.107'],\n",
    "        'ovi':[2,'Ovis_aries','Oar_v3.1.107'],\n",
    "        'gal':[3,'Gallus_gallus','gca000002315v5.GRCg6a.107'],\n",
    "        'sus':[4,'Sus_scrofa','Sscrofa11.1.107'],\n",
    "        'equ':[5,'Equus_caballus','EquCab3.0.107'],}  \n",
    "#\n",
    "all_species_table = pd.DataFrame([\n",
    "    ['bta',0,'Bos_taurus','ARS-UCD1.2.107'],\n",
    "    ['cap',1,'Capra_hircus','ARS1.107'],\n",
    "    ['ovi',2,'Ovis_aries','Oar_v3.1.107'],\n",
    "    ['gal', 3,'Gallus_gallus','gca000002315v5.GRCg6a.107'],\n",
    "    ['sus', 4,'Sus_scrofa','Sscrofa11.1.107'],\n",
    "    ['equ',5,'Equus_caballus','EquCab3.0.107'],\n",
    "])\n",
    "all_species_table.columns = ['name_short','ek_species','name_long','current_gtf']\n",
    "all_species_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### id mapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "BASE = '/Users/liulihe95/EnrichKitDB'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ek_gene_id</th>\n",
       "      <th>species</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ensembl_symbol</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>ncbi_symbol</th>\n",
       "      <th>vgnc_id</th>\n",
       "      <th>vgnc_symbol</th>\n",
       "      <th>hgnc_orthologs</th>\n",
       "      <th>human_gene_id</th>\n",
       "      <th>human_entrez_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSBTAG00000000005</td>\n",
       "      <td>GRK3</td>\n",
       "      <td>282136</td>\n",
       "      <td>GRK3</td>\n",
       "      <td>VGNC:53904</td>\n",
       "      <td>GRK3</td>\n",
       "      <td>HGNC:290</td>\n",
       "      <td>ENSG00000100077</td>\n",
       "      <td>157.0</td>\n",
       "      <td>GRK3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSBTAG00000000008</td>\n",
       "      <td>KCNJ1</td>\n",
       "      <td>539250</td>\n",
       "      <td>KCNJ1</td>\n",
       "      <td>VGNC:30453</td>\n",
       "      <td>KCNJ1</td>\n",
       "      <td>HGNC:6255</td>\n",
       "      <td>ENSG00000151704</td>\n",
       "      <td>3758.0</td>\n",
       "      <td>KCNJ1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSBTAG00000000009</td>\n",
       "      <td>FOXF1</td>\n",
       "      <td>524813</td>\n",
       "      <td>FOXF1</td>\n",
       "      <td>VGNC:29084</td>\n",
       "      <td>FOXF1</td>\n",
       "      <td>HGNC:3809</td>\n",
       "      <td>ENSG00000103241</td>\n",
       "      <td>2294.0</td>\n",
       "      <td>FOXF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSBTAG00000000010</td>\n",
       "      <td>UBL7</td>\n",
       "      <td>511289</td>\n",
       "      <td>UBL7</td>\n",
       "      <td>VGNC:50128</td>\n",
       "      <td>UBL7</td>\n",
       "      <td>HGNC:28221</td>\n",
       "      <td>ENSG00000138629</td>\n",
       "      <td>84993.0</td>\n",
       "      <td>UBL7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSBTAG00000000011</td>\n",
       "      <td>TDH</td>\n",
       "      <td>511957</td>\n",
       "      <td>TDH</td>\n",
       "      <td>VGNC:108945</td>\n",
       "      <td>TDH</td>\n",
       "      <td>HGNC:15547</td>\n",
       "      <td>ENSG00000154316</td>\n",
       "      <td>157739</td>\n",
       "      <td>TDH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328425</th>\n",
       "      <td>328425</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.14484e+08</td>\n",
       "      <td>MIR6529</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328426</th>\n",
       "      <td>328426</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.14484e+08</td>\n",
       "      <td>MIR10389</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328427</th>\n",
       "      <td>328427</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.14484e+08</td>\n",
       "      <td>MIR6516</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328428</th>\n",
       "      <td>328428</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.14484e+08</td>\n",
       "      <td>MIR33B</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328429</th>\n",
       "      <td>328429</td>\n",
       "      <td>5</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.16622e+08</td>\n",
       "      <td>DERPC</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>328430 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ek_gene_id  species             gene_id ensembl_symbol    entrez_id  \\\n",
       "0                0        0  ENSBTAG00000000005           GRK3       282136   \n",
       "1                1        0  ENSBTAG00000000008          KCNJ1       539250   \n",
       "2                2        0  ENSBTAG00000000009          FOXF1       524813   \n",
       "3                3        0  ENSBTAG00000000010           UBL7       511289   \n",
       "4                4        0  ENSBTAG00000000011            TDH       511957   \n",
       "...            ...      ...                 ...            ...          ...   \n",
       "328425      328425        5                                     1.14484e+08   \n",
       "328426      328426        5                                     1.14484e+08   \n",
       "328427      328427        5                                     1.14484e+08   \n",
       "328428      328428        5                                     1.14484e+08   \n",
       "328429      328429        5                                     1.16622e+08   \n",
       "\n",
       "       ncbi_symbol      vgnc_id vgnc_symbol hgnc_orthologs    human_gene_id  \\\n",
       "0             GRK3   VGNC:53904        GRK3       HGNC:290  ENSG00000100077   \n",
       "1            KCNJ1   VGNC:30453       KCNJ1      HGNC:6255  ENSG00000151704   \n",
       "2            FOXF1   VGNC:29084       FOXF1      HGNC:3809  ENSG00000103241   \n",
       "3             UBL7   VGNC:50128        UBL7     HGNC:28221  ENSG00000138629   \n",
       "4              TDH  VGNC:108945         TDH     HGNC:15547  ENSG00000154316   \n",
       "...            ...          ...         ...            ...              ...   \n",
       "328425     MIR6529                                                            \n",
       "328426    MIR10389                                                            \n",
       "328427     MIR6516                                                            \n",
       "328428      MIR33B                                                            \n",
       "328429       DERPC                                                            \n",
       "\n",
       "       human_entrez_id hgnc_symbol  \n",
       "0                157.0        GRK3  \n",
       "1               3758.0       KCNJ1  \n",
       "2               2294.0       FOXF1  \n",
       "3              84993.0        UBL7  \n",
       "4               157739         TDH  \n",
       "...                ...         ...  \n",
       "328425                              \n",
       "328426                              \n",
       "328427                              \n",
       "328428                              \n",
       "328429                              \n",
       "\n",
       "[328430 rows x 12 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "input_path = os.path.join(BASE,'data','tmp','id_mapper')\n",
    "id_mapper_all_updated = pd.read_csv(os.path.join(input_path,'id_mapper_all_updated.txt'),low_memory=False)\n",
    "id_mapper_all_updated.pop('Unnamed: 0')\n",
    "id_mapper_all_updated.fillna('')\n",
    "id_mapper_all_updated = id_mapper_all_updated.sort_values(by=['species','gene_id'])\n",
    "\n",
    "id_mapper_all_updated.reset_index(drop=True, inplace=True)\n",
    "id_mapper_all_updated['ek_gene_id'] = id_mapper_all_updated.index\n",
    "first_column = id_mapper_all_updated.pop('ek_gene_id')\n",
    "id_mapper_all_updated.insert(0, 'ek_gene_id', first_column)\n",
    "id_mapper_all_updated.replace(np.nan,'', regex = True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "float"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(id_mapper_all_updated.iloc[328429]['gene_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 1, 2, 3, 4, 5])"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_mapper_all_updated.species.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ek_gene_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSBTAG00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSBTAG00000000008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSBTAG00000000009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSBTAG00000000010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSBTAG00000000011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300735</th>\n",
       "      <td>300735</td>\n",
       "      <td>ENSSSCG00000063551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300736</th>\n",
       "      <td>300736</td>\n",
       "      <td>ENSSSCG00000063552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300737</th>\n",
       "      <td>300737</td>\n",
       "      <td>ENSSSCG00000063553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300738</th>\n",
       "      <td>300738</td>\n",
       "      <td>ENSSSCG00000063554</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300739</th>\n",
       "      <td>300739</td>\n",
       "      <td>ENSSSCG00000063555</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>194015 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ek_gene_id             gene_id\n",
       "0                0  ENSBTAG00000000005\n",
       "1                1  ENSBTAG00000000008\n",
       "2                2  ENSBTAG00000000009\n",
       "3                3  ENSBTAG00000000010\n",
       "4                4  ENSBTAG00000000011\n",
       "...            ...                 ...\n",
       "300735      300735  ENSSSCG00000063551\n",
       "300736      300736  ENSSSCG00000063552\n",
       "300737      300737  ENSSSCG00000063553\n",
       "300738      300738  ENSSSCG00000063554\n",
       "300739      300739  ENSSSCG00000063555\n",
       "\n",
       "[194015 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_mapper_tiny_no_dup = id_mapper_all_updated[['ek_gene_id','gene_id']].drop_duplicates().dropna()\n",
    "id_mapper_tiny_no_dup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ek_gene_id</th>\n",
       "      <th>species</th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_biotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17</td>\n",
       "      <td>65389743</td>\n",
       "      <td>65505336</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>29</td>\n",
       "      <td>32214439</td>\n",
       "      <td>32244810</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>12338037</td>\n",
       "      <td>12342272</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>21</td>\n",
       "      <td>34209956</td>\n",
       "      <td>34223394</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>7950815</td>\n",
       "      <td>7971600</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172977</th>\n",
       "      <td>132848</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>24659939</td>\n",
       "      <td>24678092</td>\n",
       "      <td>+</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172978</th>\n",
       "      <td>132849</td>\n",
       "      <td>5</td>\n",
       "      <td>22</td>\n",
       "      <td>49954554</td>\n",
       "      <td>49956116</td>\n",
       "      <td>+</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172979</th>\n",
       "      <td>132850</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "      <td>43554850</td>\n",
       "      <td>43558334</td>\n",
       "      <td>+</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172980</th>\n",
       "      <td>132851</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>43781991</td>\n",
       "      <td>43782371</td>\n",
       "      <td>+</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>172981</th>\n",
       "      <td>132852</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>16780475</td>\n",
       "      <td>16784050</td>\n",
       "      <td>+</td>\n",
       "      <td>lncRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>172982 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ek_gene_id  species  seqname     start       end strand  \\\n",
       "0                0        0       17  65389743  65505336      +   \n",
       "1                1        0       29  32214439  32244810      -   \n",
       "2                2        0       18  12338037  12342272      +   \n",
       "3                3        0       21  34209956  34223394      +   \n",
       "4                4        0        8   7950815   7971600      -   \n",
       "...            ...      ...      ...       ...       ...    ...   \n",
       "172977      132848        5       19  24659939  24678092      +   \n",
       "172978      132849        5       22  49954554  49956116      +   \n",
       "172979      132850        5        9  43554850  43558334      +   \n",
       "172980      132851        5        5  43781991  43782371      +   \n",
       "172981      132852        5        5  16780475  16784050      +   \n",
       "\n",
       "          gene_biotype  \n",
       "0       protein_coding  \n",
       "1       protein_coding  \n",
       "2       protein_coding  \n",
       "3       protein_coding  \n",
       "4       protein_coding  \n",
       "...                ...  \n",
       "172977          lncRNA  \n",
       "172978          lncRNA  \n",
       "172979          lncRNA  \n",
       "172980  protein_coding  \n",
       "172981          lncRNA  \n",
       "\n",
       "[172982 rows x 7 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_path = os.path.join(BASE,'data','tmp','annotation')\n",
    "all_out = []\n",
    "for item in os.listdir(input_path):\n",
    "        if item.startswith('genes_'):\n",
    "            cur_name_short = item.split('_')[-1].split('.')[0]\n",
    "            tmp_df = pd.read_csv(os.path.join(input_path,item))\n",
    "            tmp_df.insert(0, \"species\", all_species[cur_name_short][0])\n",
    "            all_out.append(tmp_df)\n",
    "all_out_df = pd.concat(all_out)\n",
    "all_out_df = all_out_df.sort_values(by=['species','gene_id'])\n",
    "\n",
    "all_out_df_tagged = pd.merge(all_out_df, id_mapper_tiny_no_dup, how='left', on='gene_id')\n",
    "first_column = all_out_df_tagged.pop('ek_gene_id')\n",
    "all_out_df_tagged.insert(0, 'ek_gene_id', first_column)\n",
    "genes_all = all_out_df_tagged.drop(columns=['gene_id'])\n",
    "genes_all"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ek_gene_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSBTAE00000496938</td>\n",
       "      <td>65389743</td>\n",
       "      <td>65389903</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSBTAT00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSBTAE00000000019</td>\n",
       "      <td>65480158</td>\n",
       "      <td>65480252</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSBTAT00000000005+ENSBTAT00000068309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSBTAE00000240555</td>\n",
       "      <td>65416282</td>\n",
       "      <td>65416358</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSBTAT00000000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSBTAE00000000010</td>\n",
       "      <td>65442649</td>\n",
       "      <td>65442722</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSBTAT00000000005+ENSBTAT00000068309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSBTAE00000000011</td>\n",
       "      <td>65454225</td>\n",
       "      <td>65454326</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSBTAT00000000005+ENSBTAT00000068309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1542252</th>\n",
       "      <td>300736</td>\n",
       "      <td>ENSSSCE00000496980</td>\n",
       "      <td>108776332</td>\n",
       "      <td>108776643</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSSSCT00000092591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1542253</th>\n",
       "      <td>300737</td>\n",
       "      <td>ENSSSCE00000481413</td>\n",
       "      <td>14345924</td>\n",
       "      <td>14351048</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSSSCT00000097588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1542254</th>\n",
       "      <td>300738</td>\n",
       "      <td>ENSSSCE00000493407</td>\n",
       "      <td>47881118</td>\n",
       "      <td>47881348</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSSSCT00000096902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1542255</th>\n",
       "      <td>300739</td>\n",
       "      <td>ENSSSCE00000521169</td>\n",
       "      <td>32510844</td>\n",
       "      <td>32511530</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSSSCT00000103472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1542256</th>\n",
       "      <td>300739</td>\n",
       "      <td>ENSSSCE00000505115</td>\n",
       "      <td>32538368</td>\n",
       "      <td>32538460</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSSSCT00000103472</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1542257 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ek_gene_id             exon_id      start        end    gene_biotype  \\\n",
       "0                 0  ENSBTAE00000496938   65389743   65389903  protein_coding   \n",
       "22                0  ENSBTAE00000000019   65480158   65480252  protein_coding   \n",
       "21                0  ENSBTAE00000240555   65416282   65416358  protein_coding   \n",
       "20                0  ENSBTAE00000000010   65442649   65442722  protein_coding   \n",
       "19                0  ENSBTAE00000000011   65454225   65454326  protein_coding   \n",
       "...             ...                 ...        ...        ...             ...   \n",
       "1542252      300736  ENSSSCE00000496980  108776332  108776643  protein_coding   \n",
       "1542253      300737  ENSSSCE00000481413   14345924   14351048          lncRNA   \n",
       "1542254      300738  ENSSSCE00000493407   47881118   47881348  protein_coding   \n",
       "1542255      300739  ENSSSCE00000521169   32510844   32511530  protein_coding   \n",
       "1542256      300739  ENSSSCE00000505115   32538368   32538460  protein_coding   \n",
       "\n",
       "        strand                          transcript_id  \n",
       "0            +                     ENSBTAT00000000005  \n",
       "22           +  ENSBTAT00000000005+ENSBTAT00000068309  \n",
       "21           +                     ENSBTAT00000000005  \n",
       "20           +  ENSBTAT00000000005+ENSBTAT00000068309  \n",
       "19           +  ENSBTAT00000000005+ENSBTAT00000068309  \n",
       "...        ...                                    ...  \n",
       "1542252      -                     ENSSSCT00000092591  \n",
       "1542253      +                     ENSSSCT00000097588  \n",
       "1542254      -                     ENSSSCT00000096902  \n",
       "1542255      +                     ENSSSCT00000103472  \n",
       "1542256      +                     ENSSSCT00000103472  \n",
       "\n",
       "[1542257 rows x 7 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_path = os.path.join(BASE,'data','tmp','annotation')\n",
    "all_out = []\n",
    "for item in os.listdir(input_path):\n",
    "        if item.startswith('exons_'):\n",
    "            cur_name_short = item.split('_')[-1].split('.')[0]\n",
    "            tmp_df = pd.read_csv(os.path.join(input_path,item))\n",
    "#             tmp_df.insert(0, \"species\", all_species[cur_name_short][0])\n",
    "            all_out.append(tmp_df)\n",
    "all_out_df = pd.concat(all_out)\n",
    "all_out_df = all_out_df.sort_values(by=['gene_id'])\n",
    "\n",
    "all_out_df_tagged = pd.merge(all_out_df, id_mapper_tiny_no_dup, how='left', on='gene_id')\n",
    "first_column = all_out_df_tagged.pop('ek_gene_id')\n",
    "all_out_df_tagged.insert(0, 'ek_gene_id', first_column)\n",
    "exons_all = all_out_df_tagged.drop(columns=['gene_id'])\n",
    "exons_all = exons_all.sort_values(by=['ek_gene_id'])\n",
    "exons_all"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_path = os.path.join(BASE,'data','tmp','annotation')\n",
    "all_out = []\n",
    "for item in os.listdir(input_path):\n",
    "        if item.startswith('features_'):\n",
    "            cur_name_short = item.split('_')[-1].split('.')[0]\n",
    "            tmp_df = pd.read_csv(os.path.join(input_path,item))\n",
    "#             tmp_df.insert(0, \"species\", all_species[cur_name_short][0])\n",
    "            all_out.append(tmp_df)\n",
    "all_out_df = pd.concat(all_out)\n",
    "all_out_df = all_out_df.sort_values(by=['gene_id'])\n",
    "\n",
    "all_out_df_tagged = pd.merge(all_out_df, id_mapper_tiny_no_dup, how='left', on='gene_id')\n",
    "first_column = all_out_df_tagged.pop('ek_gene_id')\n",
    "all_out_df_tagged.insert(0, 'ek_gene_id', first_column)\n",
    "features_all = all_out_df_tagged.drop(columns=['gene_id'])\n",
    "features_all = features_all.sort_values(by=['ek_gene_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ek_gene_id</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>65389743</td>\n",
       "      <td>65389790</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>0</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65456261</td>\n",
       "      <td>65456335</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>0</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65462772</td>\n",
       "      <td>65462833</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>0</td>\n",
       "      <td>CDS</td>\n",
       "      <td>65465537</td>\n",
       "      <td>65465588</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>0</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>65504654</td>\n",
       "      <td>65504656</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639823</th>\n",
       "      <td>300739</td>\n",
       "      <td>CDS</td>\n",
       "      <td>32510844</td>\n",
       "      <td>32511530</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639820</th>\n",
       "      <td>300739</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>32510844</td>\n",
       "      <td>32510846</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639821</th>\n",
       "      <td>300739</td>\n",
       "      <td>CDS</td>\n",
       "      <td>32538368</td>\n",
       "      <td>32538457</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639822</th>\n",
       "      <td>300739</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>32538458</td>\n",
       "      <td>32538460</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3639824</th>\n",
       "      <td>300739</td>\n",
       "      <td>transcript</td>\n",
       "      <td>32510844</td>\n",
       "      <td>32538460</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3639825 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ek_gene_id         feature     start       end    gene_biotype strand\n",
       "0                 0  five_prime_utr  65389743  65389790  protein_coding      +\n",
       "26                0             CDS  65456261  65456335  protein_coding      +\n",
       "27                0             CDS  65462772  65462833  protein_coding      +\n",
       "28                0             CDS  65465537  65465588  protein_coding      +\n",
       "29                0      stop_codon  65504654  65504656  protein_coding      +\n",
       "...             ...             ...       ...       ...             ...    ...\n",
       "3639823      300739             CDS  32510844  32511530  protein_coding      +\n",
       "3639820      300739     start_codon  32510844  32510846  protein_coding      +\n",
       "3639821      300739             CDS  32538368  32538457  protein_coding      +\n",
       "3639822      300739      stop_codon  32538458  32538460  protein_coding      +\n",
       "3639824      300739      transcript  32510844  32538460  protein_coding      +\n",
       "\n",
       "[3639825 rows x 6 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_all"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### computed feature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ek_gene_id</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>splice acceptor</td>\n",
       "      <td>65416015</td>\n",
       "      <td>65416065</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>splice acceptor</td>\n",
       "      <td>65485580</td>\n",
       "      <td>65485630</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>splice donor</td>\n",
       "      <td>65485739</td>\n",
       "      <td>65485789</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>intron</td>\n",
       "      <td>65485790</td>\n",
       "      <td>65491986</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>splice acceptor</td>\n",
       "      <td>65491987</td>\n",
       "      <td>65492037</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3784736</th>\n",
       "      <td>300739</td>\n",
       "      <td>upstream</td>\n",
       "      <td>32500843</td>\n",
       "      <td>32510843</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3784737</th>\n",
       "      <td>300739</td>\n",
       "      <td>splice donor</td>\n",
       "      <td>32511531</td>\n",
       "      <td>32511581</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3784738</th>\n",
       "      <td>300739</td>\n",
       "      <td>intron</td>\n",
       "      <td>32511582</td>\n",
       "      <td>32538316</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3784739</th>\n",
       "      <td>300739</td>\n",
       "      <td>splice acceptor</td>\n",
       "      <td>32538317</td>\n",
       "      <td>32538367</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3784740</th>\n",
       "      <td>300739</td>\n",
       "      <td>downstream</td>\n",
       "      <td>32538461</td>\n",
       "      <td>32548461</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3784741 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ek_gene_id          feature     start       end    gene_biotype  \\\n",
       "0                 0  splice acceptor  65416015  65416065  protein_coding   \n",
       "1                 0  splice acceptor  65485580  65485630  protein_coding   \n",
       "2                 0     splice donor  65485739  65485789  protein_coding   \n",
       "3                 0           intron  65485790  65491986  protein_coding   \n",
       "4                 0  splice acceptor  65491987  65492037  protein_coding   \n",
       "...             ...              ...       ...       ...             ...   \n",
       "3784736      300739         upstream  32500843  32510843  protein_coding   \n",
       "3784737      300739     splice donor  32511531  32511581  protein_coding   \n",
       "3784738      300739           intron  32511582  32538316  protein_coding   \n",
       "3784739      300739  splice acceptor  32538317  32538367  protein_coding   \n",
       "3784740      300739       downstream  32538461  32548461  protein_coding   \n",
       "\n",
       "        strand  \n",
       "0            +  \n",
       "1            +  \n",
       "2            +  \n",
       "3            +  \n",
       "4            +  \n",
       "...        ...  \n",
       "3784736      +  \n",
       "3784737      +  \n",
       "3784738      +  \n",
       "3784739      +  \n",
       "3784740      +  \n",
       "\n",
       "[3784741 rows x 6 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_path = os.path.join(BASE,'data','tmp','annotation')\n",
    "all_out = []\n",
    "for item in os.listdir(input_path):\n",
    "        if item.startswith('computed_feature_'):\n",
    "            cur_name_short = item.split('_')[-1].split('.')[0]\n",
    "            tmp_df = pd.read_csv(os.path.join(input_path,item))\n",
    "#             tmp_df.insert(0, \"species\", all_species[cur_name_short][0])\n",
    "            all_out.append(tmp_df)\n",
    "all_out_df = pd.concat(all_out)\n",
    "all_out_df = all_out_df.sort_values(by=['gene_id'])\n",
    "\n",
    "all_out_df_tagged = pd.merge(all_out_df, id_mapper_tiny_no_dup, how='left', on='gene_id')\n",
    "first_column = all_out_df_tagged.pop('ek_gene_id')\n",
    "all_out_df_tagged.insert(0, 'ek_gene_id', first_column)\n",
    "computed_feature_all = all_out_df_tagged.drop(columns=['gene_id'])\n",
    "computed_feature_all"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pathway meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'10/03/22'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "current_time = datetime.now().strftime(\"%D\")\n",
    "current_time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ek_gene_id',\n",
       " 'species',\n",
       " 'gene_id',\n",
       " 'ensembl_symbol',\n",
       " 'entrez_id',\n",
       " 'ncbi_symbol',\n",
       " 'vgnc_id',\n",
       " 'vgnc_symbol',\n",
       " 'hgnc_orthologs',\n",
       " 'human_gene_id',\n",
       " 'human_entrez_id',\n",
       " 'hgnc_symbol']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(id_mapper_all_updated.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'models' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-16-35e066fd1844>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpathway_meta_id\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodels\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mIntegerField\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mverbose_name\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'database unit meta_data id'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mprimary_key\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mspecies\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodels\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCharField\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mverbose_name\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'current species'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_length\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodels\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCharField\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mverbose_name\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'current db unit name'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_length\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mid_type\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodels\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCharField\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mverbose_name\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'the id type of the current db unit'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_length\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m50\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mupdate_time\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmodels\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCharField\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mverbose_name\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'current db unit update time'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_length\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m20\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'models' is not defined"
     ]
    }
   ],
   "source": [
    "# pathway_meta_id = models.IntegerField(verbose_name='database unit meta_data id', primary_key=True)\n",
    "# species = models.CharField(verbose_name='current species', max_length=10)\n",
    "# name = models.CharField(verbose_name='current db unit name', max_length=20)\n",
    "# id_type = models.CharField(verbose_name='the id type of the current db unit', max_length=50)\n",
    "# update_time = models.CharField(verbose_name='current db unit update time', max_length=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_meta_id</th>\n",
       "      <th>species</th>\n",
       "      <th>name</th>\n",
       "      <th>id_type</th>\n",
       "      <th>update_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>all</td>\n",
       "      <td>msigdb</td>\n",
       "      <td>human_entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>ovi</td>\n",
       "      <td>go</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>bta</td>\n",
       "      <td>go</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>equ</td>\n",
       "      <td>go</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>gal</td>\n",
       "      <td>go</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>cap</td>\n",
       "      <td>go</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>sus</td>\n",
       "      <td>go</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>ovi</td>\n",
       "      <td>interpro</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>bta</td>\n",
       "      <td>interpro</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>equ</td>\n",
       "      <td>interpro</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>gal</td>\n",
       "      <td>interpro</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>cap</td>\n",
       "      <td>interpro</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>sus</td>\n",
       "      <td>interpro</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>gal</td>\n",
       "      <td>kegg</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>cap</td>\n",
       "      <td>kegg</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>bta</td>\n",
       "      <td>kegg</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>sus</td>\n",
       "      <td>kegg</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>ovi</td>\n",
       "      <td>kegg</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>equ</td>\n",
       "      <td>kegg</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>bta</td>\n",
       "      <td>mesh</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>equ</td>\n",
       "      <td>mesh</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>gal</td>\n",
       "      <td>mesh</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>23</td>\n",
       "      <td>ssc</td>\n",
       "      <td>mesh</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>24</td>\n",
       "      <td>bta</td>\n",
       "      <td>reactome</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25</td>\n",
       "      <td>gal</td>\n",
       "      <td>reactome</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>26</td>\n",
       "      <td>sus</td>\n",
       "      <td>reactome</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    pathway_meta_id species      name          id_type update_time\n",
       "0                 1     all    msigdb  human_entrez_id    10/03/22\n",
       "1                 2     ovi        go          gene_id    10/03/22\n",
       "2                 3     bta        go          gene_id    10/03/22\n",
       "3                 4     equ        go          gene_id    10/03/22\n",
       "4                 5     gal        go          gene_id    10/03/22\n",
       "5                 6     cap        go          gene_id    10/03/22\n",
       "6                 7     sus        go          gene_id    10/03/22\n",
       "7                 8     ovi  interpro          gene_id    10/03/22\n",
       "8                 9     bta  interpro          gene_id    10/03/22\n",
       "9                10     equ  interpro          gene_id    10/03/22\n",
       "10               11     gal  interpro          gene_id    10/03/22\n",
       "11               12     cap  interpro          gene_id    10/03/22\n",
       "12               13     sus  interpro          gene_id    10/03/22\n",
       "13               14     gal      kegg        entrez_id    10/03/22\n",
       "14               15     cap      kegg        entrez_id    10/03/22\n",
       "15               16     bta      kegg        entrez_id    10/03/22\n",
       "16               17     sus      kegg        entrez_id    10/03/22\n",
       "17               18     ovi      kegg        entrez_id    10/03/22\n",
       "18               19     equ      kegg        entrez_id    10/03/22\n",
       "19               20     bta      mesh        entrez_id    10/03/22\n",
       "20               21     equ      mesh        entrez_id    10/03/22\n",
       "21               22     gal      mesh        entrez_id    10/03/22\n",
       "22               23     ssc      mesh        entrez_id    10/03/22\n",
       "23               24     bta  reactome        entrez_id    10/03/22\n",
       "24               25     gal  reactome        entrez_id    10/03/22\n",
       "25               26     sus  reactome        entrez_id    10/03/22"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_db_names = ['msigdb', 'go','interpro', 'kegg','mesh','reactome']\n",
    "all_species_names = list(all_species.keys())\n",
    "all_db_id_type = {\n",
    "    'msigdb': 'human_entrez_id',\n",
    "    'go': 'gene_id',\n",
    "    'interpro': 'gene_id',\n",
    "    'kegg': 'entrez_id',\n",
    "    'mesh': 'entrez_id',\n",
    "    'reactome':'entrez_id',\n",
    "}\n",
    "out = []\n",
    "for db in all_db_names:\n",
    "    pathways_dir_path = os.path.join(BASE,'data','tmp', db)\n",
    "    involve_list = os.listdir(pathways_dir_path)\n",
    "    if db == 'msigdb':\n",
    "        tmp_tuple = ['all', db, all_db_id_type[db], current_time]\n",
    "        out.append(tmp_tuple)\n",
    "    else:\n",
    "        for species_involve in involve_list:\n",
    "            if species_involve.endswith('_involve.txt'):\n",
    "                tmp_tuple = [species_involve.split('_')[0],db, all_db_id_type[db], current_time]\n",
    "                out.append(tmp_tuple)             \n",
    "pathway_meta = pd.DataFrame(out)\n",
    "pathway_meta.columns = ['species','name','id_type','update_time']\n",
    "pathway_meta.reset_index(drop=True, inplace=True)\n",
    "pathway_meta.insert(0, 'pathway_meta_id', pathway_meta.index + 1)\n",
    "pathway_meta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ek_pathway_id</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>pathway_description</th>\n",
       "      <th>pathway_meta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>M12793</td>\n",
       "      <td>Ensembl 103 genes in cytogenetic band chr1p11</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>M14892</td>\n",
       "      <td>Ensembl 103 genes in cytogenetic band chr1p12</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>M6974</td>\n",
       "      <td>Ensembl 103 genes in cytogenetic band chr1p13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>M16131</td>\n",
       "      <td>Ensembl 103 genes in cytogenetic band chr1p21</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>M15893</td>\n",
       "      <td>Ensembl 103 genes in cytogenetic band chr1p22</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278137</th>\n",
       "      <td>278137</td>\n",
       "      <td>R-SSC-380612</td>\n",
       "      <td>Metabolism of serotonin</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278138</th>\n",
       "      <td>278138</td>\n",
       "      <td>R-SSC-380615</td>\n",
       "      <td>Serotonin clearance from the synaptic cleft</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278139</th>\n",
       "      <td>278139</td>\n",
       "      <td>R-SSC-9617629</td>\n",
       "      <td>Regulation of FOXO transcriptional activity by...</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278140</th>\n",
       "      <td>278140</td>\n",
       "      <td>R-SSC-9706374</td>\n",
       "      <td>FLT3 signaling through SRC family kinases</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>278141</th>\n",
       "      <td>278141</td>\n",
       "      <td>R-SSC-8866376</td>\n",
       "      <td>Reelin signalling pathway</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>278142 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ek_pathway_id     pathway_id  \\\n",
       "0                   0         M12793   \n",
       "1                   1         M14892   \n",
       "2                   2          M6974   \n",
       "3                   3         M16131   \n",
       "4                   4         M15893   \n",
       "...               ...            ...   \n",
       "278137         278137   R-SSC-380612   \n",
       "278138         278138   R-SSC-380615   \n",
       "278139         278139  R-SSC-9617629   \n",
       "278140         278140  R-SSC-9706374   \n",
       "278141         278141  R-SSC-8866376   \n",
       "\n",
       "                                      pathway_description  pathway_meta  \n",
       "0           Ensembl 103 genes in cytogenetic band chr1p11             1  \n",
       "1           Ensembl 103 genes in cytogenetic band chr1p12             1  \n",
       "2           Ensembl 103 genes in cytogenetic band chr1p13             1  \n",
       "3           Ensembl 103 genes in cytogenetic band chr1p21             1  \n",
       "4           Ensembl 103 genes in cytogenetic band chr1p22             1  \n",
       "...                                                   ...           ...  \n",
       "278137                            Metabolism of serotonin            26  \n",
       "278138        Serotonin clearance from the synaptic cleft            26  \n",
       "278139  Regulation of FOXO transcriptional activity by...            26  \n",
       "278140          FLT3 signaling through SRC family kinases            26  \n",
       "278141                          Reelin signalling pathway            26  \n",
       "\n",
       "[278142 rows x 4 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_db_names = ['msigdb', 'go','interpro', 'kegg','mesh','reactome']\n",
    "all_species_names = list(all_species.keys())\n",
    "all_db_id_type = {\n",
    "    'msigdb': 'human_entrez_id',\n",
    "    'go': 'gene_id',\n",
    "    'interpro': 'gene_id',\n",
    "    'kegg': 'entrez_id',\n",
    "    'mesh': 'entrez_id',\n",
    "    'reactome':'entrez_id',\n",
    "}\n",
    "out = []\n",
    "for db in all_db_names:\n",
    "    pathways_dir_path = os.path.join(BASE,'data','tmp', db)\n",
    "    pathway_description_list = os.listdir(pathways_dir_path)        \n",
    "    if db == 'msigdb':\n",
    "        for description in pathway_description_list:\n",
    "            if description.endswith('pathway_list.txt'):\n",
    "                tmp_df = pd.read_csv(os.path.join(pathways_dir_path, description))\n",
    "                tmp_df.insert(2,'pathway_meta',1)\n",
    "                out.append(tmp_df)\n",
    "\n",
    "        next\n",
    "    else:\n",
    "        for description in pathway_description_list:\n",
    "            if description.endswith('_description.txt'):\n",
    "                tmp_db = db\n",
    "                tmp_species = description.split('_')[0]\n",
    "                if tmp_db == 'mesh' and tmp_species == 'all':\n",
    "                    tmp_id_list = pathway_meta.loc[pathway_meta.name == db]['pathway_meta_id']\n",
    "                    for id_mesh in tmp_id_list:\n",
    "                        tmp_df = pd.read_csv(os.path.join(pathways_dir_path,description))                    \n",
    "                        tmp_df.insert(2,'pathway_meta',id_mesh)\n",
    "                        out.append(tmp_df)\n",
    "                else:  \n",
    "                    tmp_id = pathway_meta.loc[(pathway_meta.species == tmp_species) & (pathway_meta.name == db)]['pathway_meta_id']\n",
    "                    tmp_df = pd.read_csv(os.path.join(pathways_dir_path,description))\n",
    "                    tmp_df.insert(2,'pathway_meta',tmp_id.values[0])\n",
    "                    out.append(tmp_df)                             \n",
    "            \n",
    "all_pathway = pd.concat(out)\n",
    "all_pathway.reset_index(drop=True, inplace=True)\n",
    "all_pathway.insert(0, 'ek_pathway_id', all_pathway.index)\n",
    "all_pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 1,  6,  2,  5,  4,  3,  7, 12,  8, 11, 10,  9, 13, 15, 18, 14, 19,\n",
       "       16, 17, 20, 21, 22, 23, 25, 24, 26])"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_pathway['pathway_meta'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### involve"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['ek_pathway_id', 'pathway_id', 'pathway_description', 'pathway_meta'], dtype='object')"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_pathway.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pathway_meta_id     int64\n",
       "species            object\n",
       "name               object\n",
       "id_type            object\n",
       "update_time        object\n",
       "dtype: object"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_meta.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'id_mapper_all_updated' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-6723b6b66677>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0mid_mapper_all_updated\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mid_mapper_all_updated\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mdb\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mall_db_names\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mpathways_dir_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mBASE\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'data'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'tmp'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0mpathway_involve_list\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlistdir\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpathways_dir_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'id_mapper_all_updated' is not defined"
     ]
    }
   ],
   "source": [
    "out = []\n",
    "id_mapper_all_updated = id_mapper_all_updated.astype(str)\n",
    "for db in all_db_names:\n",
    "    pathways_dir_path = os.path.join(BASE,'data','tmp', db)\n",
    "    pathway_involve_list = os.listdir(pathways_dir_path)        \n",
    "    if db == 'msigdb':\n",
    "        for involve in pathway_involve_list:\n",
    "            if involve.endswith('involve_list.txt'):\n",
    "                print('currently compose involve list for msigdb')\n",
    "                tmp_df = pd.read_csv(os.path.join(pathways_dir_path,involve))\n",
    "                tmp_df = tmp_df.astype({'gene_id': float})\n",
    "                tmp_df = tmp_df.dropna().astype({'gene_id': int})\n",
    "                tmp_df = tmp_df.astype({'gene_id': str})\n",
    "                \n",
    "                target_pathway_meta = 1\n",
    "        \n",
    "                pathway_map_tmp = all_pathway[all_pathway['pathway_meta'] == target_pathway_meta]\n",
    "        \n",
    "                tmp_df_2 = pd.merge(tmp_df, pathway_map_tmp[['ek_pathway_id', 'pathway_id']],\n",
    "                                    how = 'left', \n",
    "                                    on='pathway_id')\n",
    "                out_tmp = tmp_df_2[['gene_id','ek_pathway_id']]\n",
    "                out.append(out_tmp)\n",
    "        next\n",
    "    else:\n",
    "        for involve in pathway_involve_list:\n",
    "            if involve.endswith('involve.txt'):\n",
    "                tmp_db = db\n",
    "                tmp_species = involve.split('_')[0]\n",
    "                print('currently compose involve list of ', tmp_db, ' database for species ', tmp_species)\n",
    "                tmp_df = pd.read_csv(os.path.join(pathways_dir_path,involve)).drop_duplicates()\n",
    "                tmp_df = tmp_df.astype(str)\n",
    "                \n",
    "                id_type = pathway_meta.loc[(pathway_meta.name == tmp_db) & (pathway_meta.species == tmp_species)]['id_type'].values[0]\n",
    "                \n",
    "                mapper_tmp  = id_mapper_all_updated[['ek_gene_id',id_type]].dropna().drop_duplicates()\n",
    "                if id_type == 'entrez_id':\n",
    "                    mapper_tmp = mapper_tmp.astype({id_type: float})\n",
    "                    mapper_tmp = mapper_tmp.dropna().astype({id_type: int})\n",
    "                    mapper_tmp = mapper_tmp.astype({id_type: str})\n",
    "                    \n",
    "                tmp_df_1 = pd.merge(tmp_df, mapper_tmp,\n",
    "                                    how = 'left', \n",
    "                                    left_on='gene_id', right_on = id_type).dropna()\n",
    "                \n",
    "                target_pathway_meta = pathway_meta.loc[(pathway_meta.name == tmp_db) & (pathway_meta.species == tmp_species)]['pathway_meta_id'].values[0]\n",
    "                pathway_map_tmp = all_pathway[all_pathway['pathway_meta'] == target_pathway_meta]\n",
    "                \n",
    "                tmp_df_2 = pd.merge(tmp_df_1, pathway_map_tmp[['ek_pathway_id', 'pathway_id']],\n",
    "                                    how = 'left', \n",
    "                                    on='pathway_id')#.dropna()\n",
    "                \n",
    "                out_tmp = tmp_df_2[['ek_gene_id','ek_pathway_id']]\n",
    "                out.append(out_tmp)\n",
    "print('Done!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'out' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-9bc27bdc8279>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mout\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'out' is not defined"
     ]
    }
   ],
   "source": [
    "out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>pathway_id</th>\n",
       "      <th>ek_gene_id</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>ek_pathway_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100037269</td>\n",
       "      <td>R-SSC-382551</td>\n",
       "      <td>273382</td>\n",
       "      <td>100037269</td>\n",
       "      <td>276640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100037269</td>\n",
       "      <td>R-SSC-425397</td>\n",
       "      <td>273382</td>\n",
       "      <td>100037269</td>\n",
       "      <td>276641</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>100037269</td>\n",
       "      <td>R-SSC-425407</td>\n",
       "      <td>273382</td>\n",
       "      <td>100037269</td>\n",
       "      <td>276642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>100037269</td>\n",
       "      <td>R-SSC-804914</td>\n",
       "      <td>273382</td>\n",
       "      <td>100037269</td>\n",
       "      <td>276643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>100037275</td>\n",
       "      <td>R-SSC-381426</td>\n",
       "      <td>270594</td>\n",
       "      <td>100037275</td>\n",
       "      <td>276644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22992</th>\n",
       "      <td>808512</td>\n",
       "      <td>R-SSC-6799198</td>\n",
       "      <td>275785</td>\n",
       "      <td>808512</td>\n",
       "      <td>276707</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22993</th>\n",
       "      <td>808513</td>\n",
       "      <td>R-SSC-1428517</td>\n",
       "      <td>275787</td>\n",
       "      <td>808513</td>\n",
       "      <td>276686</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22994</th>\n",
       "      <td>808513</td>\n",
       "      <td>R-SSC-1430728</td>\n",
       "      <td>275787</td>\n",
       "      <td>808513</td>\n",
       "      <td>276657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22995</th>\n",
       "      <td>808513</td>\n",
       "      <td>R-SSC-163200</td>\n",
       "      <td>275787</td>\n",
       "      <td>808513</td>\n",
       "      <td>276688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22996</th>\n",
       "      <td>808513</td>\n",
       "      <td>R-SSC-611105</td>\n",
       "      <td>275787</td>\n",
       "      <td>808513</td>\n",
       "      <td>276701</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22997 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_id     pathway_id ek_gene_id  entrez_id  ek_pathway_id\n",
       "0      100037269   R-SSC-382551     273382  100037269         276640\n",
       "1      100037269   R-SSC-425397     273382  100037269         276641\n",
       "2      100037269   R-SSC-425407     273382  100037269         276642\n",
       "3      100037269   R-SSC-804914     273382  100037269         276643\n",
       "4      100037275   R-SSC-381426     270594  100037275         276644\n",
       "...          ...            ...        ...        ...            ...\n",
       "22992     808512  R-SSC-6799198     275785     808512         276707\n",
       "22993     808513  R-SSC-1428517     275787     808513         276686\n",
       "22994     808513  R-SSC-1430728     275787     808513         276657\n",
       "22995     808513   R-SSC-163200     275787     808513         276688\n",
       "22996     808513   R-SSC-611105     275787     808513         276701\n",
       "\n",
       "[22997 rows x 5 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp_df_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pathway_meta_id</th>\n",
       "      <th>species</th>\n",
       "      <th>name</th>\n",
       "      <th>id_type</th>\n",
       "      <th>update_time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>all</td>\n",
       "      <td>msigdb</td>\n",
       "      <td>human_entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>ovi</td>\n",
       "      <td>go</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>bta</td>\n",
       "      <td>go</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>equ</td>\n",
       "      <td>go</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>gal</td>\n",
       "      <td>go</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>cap</td>\n",
       "      <td>go</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>sus</td>\n",
       "      <td>go</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>8</td>\n",
       "      <td>ovi</td>\n",
       "      <td>interpro</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9</td>\n",
       "      <td>bta</td>\n",
       "      <td>interpro</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>10</td>\n",
       "      <td>equ</td>\n",
       "      <td>interpro</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>11</td>\n",
       "      <td>gal</td>\n",
       "      <td>interpro</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12</td>\n",
       "      <td>cap</td>\n",
       "      <td>interpro</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>13</td>\n",
       "      <td>sus</td>\n",
       "      <td>interpro</td>\n",
       "      <td>gene_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>14</td>\n",
       "      <td>gal</td>\n",
       "      <td>kegg</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>15</td>\n",
       "      <td>cap</td>\n",
       "      <td>kegg</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>16</td>\n",
       "      <td>bta</td>\n",
       "      <td>kegg</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>17</td>\n",
       "      <td>sus</td>\n",
       "      <td>kegg</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>18</td>\n",
       "      <td>ovi</td>\n",
       "      <td>kegg</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>19</td>\n",
       "      <td>equ</td>\n",
       "      <td>kegg</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>20</td>\n",
       "      <td>bta</td>\n",
       "      <td>mesh</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>21</td>\n",
       "      <td>equ</td>\n",
       "      <td>mesh</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>22</td>\n",
       "      <td>gal</td>\n",
       "      <td>mesh</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>23</td>\n",
       "      <td>ssc</td>\n",
       "      <td>mesh</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>24</td>\n",
       "      <td>bta</td>\n",
       "      <td>reactome</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>25</td>\n",
       "      <td>gal</td>\n",
       "      <td>reactome</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>26</td>\n",
       "      <td>sus</td>\n",
       "      <td>reactome</td>\n",
       "      <td>entrez_id</td>\n",
       "      <td>10/03/22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    pathway_meta_id species      name          id_type update_time\n",
       "0                 1     all    msigdb  human_entrez_id    10/03/22\n",
       "1                 2     ovi        go          gene_id    10/03/22\n",
       "2                 3     bta        go          gene_id    10/03/22\n",
       "3                 4     equ        go          gene_id    10/03/22\n",
       "4                 5     gal        go          gene_id    10/03/22\n",
       "5                 6     cap        go          gene_id    10/03/22\n",
       "6                 7     sus        go          gene_id    10/03/22\n",
       "7                 8     ovi  interpro          gene_id    10/03/22\n",
       "8                 9     bta  interpro          gene_id    10/03/22\n",
       "9                10     equ  interpro          gene_id    10/03/22\n",
       "10               11     gal  interpro          gene_id    10/03/22\n",
       "11               12     cap  interpro          gene_id    10/03/22\n",
       "12               13     sus  interpro          gene_id    10/03/22\n",
       "13               14     gal      kegg        entrez_id    10/03/22\n",
       "14               15     cap      kegg        entrez_id    10/03/22\n",
       "15               16     bta      kegg        entrez_id    10/03/22\n",
       "16               17     sus      kegg        entrez_id    10/03/22\n",
       "17               18     ovi      kegg        entrez_id    10/03/22\n",
       "18               19     equ      kegg        entrez_id    10/03/22\n",
       "19               20     bta      mesh        entrez_id    10/03/22\n",
       "20               21     equ      mesh        entrez_id    10/03/22\n",
       "21               22     gal      mesh        entrez_id    10/03/22\n",
       "22               23     ssc      mesh        entrez_id    10/03/22\n",
       "23               24     bta  reactome        entrez_id    10/03/22\n",
       "24               25     gal  reactome        entrez_id    10/03/22\n",
       "25               26     sus  reactome        entrez_id    10/03/22"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
