{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import requests, sys\n",
    "import numpy as np\n",
    "from datetime import datetime\n",
    "from collections import defaultdict\n",
    "from zipfile import ZipFile\n",
    "import gzip\n",
    "import xmltodict\n",
    "import lxml\n",
    "from lxml import etree\n",
    "from io import StringIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_ensembl_symbol_helper(gene_id):\n",
    "    print('checking ens symbol for ', gene_id)\n",
    "    server = \"https://rest.ensembl.org\"\n",
    "    ext = ''.join(['/xrefs/id/', gene_id, '?'])\n",
    "\n",
    "    r = requests.get(server + ext, headers={\"Content-Type\": \"application/json\"})\n",
    "\n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "        return None\n",
    "\n",
    "    decoded = r.json()\n",
    "\n",
    "    symbol_tmp = decoded[0]['display_id']\n",
    "\n",
    "    if symbol_tmp is not None and symbol_tmp != gene_id:\n",
    "        return symbol_tmp\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "def from_ens_2_entrez(gene_id):\n",
    "    print('checking ncbi entrez id and symbol for ', gene_id)\n",
    "    target_url = 'https://www.ncbi.nlm.nih.gov/gene/?term=' + gene_id\n",
    "    response = requests.get(target_url)\n",
    "    dict_data = xmltodict.parse(response.content)\n",
    "    package = dict_data['html']['body']['div']['div']['form']['div'][0]['div'][3]['div']['div'][5]\n",
    "    entrez_id = package['div'][0]['div'][0]['span']['#text'].split(' ')[2][:-1]\n",
    "    name = package['div'][1]['div'][0]['div']['div']['dl']['dd'][0]['#text']\n",
    "    return entrez_id, name\n",
    "\n",
    "\n",
    "def get_human_orth_ncbi(gene_id):\n",
    "    print('checking ncbi human orthlogus for ', gene_id)\n",
    "    target_url = 'https://www.ncbi.nlm.nih.gov/gene/?term=' + gene_id\n",
    "    response = requests.get(target_url)\n",
    "    dict_data = xmltodict.parse(response.content)\n",
    "    package = dict_data['html']['body']['div']['div']['form']['div'][0]['div'][3]['div']['div'][5]\n",
    "    for item in package['div'][1]['div'][0]['div']['div']['dl']['dd'][7]['a']:\n",
    "        if item['#text'] == 'human':\n",
    "            human_entrez_id = item['@href'].split('/')[-1]\n",
    "    return human_entrez_id\n",
    "\n",
    "\n",
    "def parse_human_url(human_entrez_id):\n",
    "    print('parsing ncbi human orthlogus for', human_entrez_id)\n",
    "    target_url = 'https://www.ncbi.nlm.nih.gov/gene/' + human_entrez_id\n",
    "    response = requests.get(target_url)\n",
    "    dict_data = xmltodict.parse(response.content)\n",
    "    package = dict_data['html']['body']['div']['div']['form']['div'][0]['div'][3]['div']['div'][5]['div'][1]['div'][0]\n",
    "\n",
    "    human_symbol = package['div']['div']['dl']['dd'][0]['#text']\n",
    "    hgnc = package['div']['div']['dl']['dd'][2]['a']['#text']\n",
    "    ens_id = package['div']['div']['dl']['dd'][3]['a'][0]['#text']\n",
    "    if len(ens_id.split(':')) == 2:\n",
    "        ens_id = ens_id.split(':')[1]\n",
    "\n",
    "    return human_symbol, hgnc, ens_id\n",
    "\n",
    "\n",
    "def it_child(node):\n",
    "    for element in node.iterchildren():\n",
    "        if element.tag == 'a' and 'href' in element.attrib and element.attrib['href'].startswith('http://www.ncbi.nlm.nih.gov/entrez'):\n",
    "            return element.text\n",
    "        it_child(element)\n",
    "        \n",
    "def from_ens_2_entrez_use_enblorg(tmp_gene_id):\n",
    "    target_url = 'https://uswest.ensembl.org/Gene/Summary?g=' + tmp_gene_id\n",
    "    parser = etree.HTMLParser(recover=True)\n",
    "    page = requests.get(target_url)\n",
    "    html = page.content.decode(\"utf-8\")\n",
    "    tree = etree.parse(StringIO(html), parser=parser)\n",
    "    root = tree.getroot()\n",
    "    return it_child(root)\n",
    "\n",
    "\n",
    "def check_human_gene_id(gene_id):\n",
    "    server = \"https://rest.ensembl.org\"\n",
    "    ext = \"/homology/id/\" + gene_id + \"?type=orthologues;format=condensed\"\n",
    "    r = requests.get(server + ext, headers={\"Content-Type\": \"application/json\"})\n",
    "\n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "\n",
    "    decoded = r.json()\n",
    "    return decoded\n",
    "\n",
    "\n",
    "def check_ensembl_symbol_helper(gene_id):\n",
    "    print('checking ensembl symbol for', gene_id)\n",
    "    try:\n",
    "        server = \"https://rest.ensembl.org\"\n",
    "        ext = ''.join(['/xrefs/id/', gene_id, '?'])\n",
    "\n",
    "        r = requests.get(server + ext, headers={\"Content-Type\": \"application/json\"})\n",
    "\n",
    "        if not r.ok:\n",
    "            r.raise_for_status()\n",
    "            sys.exit()\n",
    "\n",
    "        decoded = r.json()\n",
    "\n",
    "        symbol_tmp = decoded[0]['display_id']\n",
    "\n",
    "        if symbol_tmp is not None and symbol_tmp != gene_id:\n",
    "            return symbol_tmp\n",
    "        else:\n",
    "            return None\n",
    "    except Exception:\n",
    "        return None\n",
    "def from_hgncid_to_hgncsymb(hgnc_id):\n",
    "    server = 'https://rest.genenames.org/'\n",
    "    ext = 'search/hgnc_id/' + str(hgnc_id.split(\":\")[-1])\n",
    "    \n",
    "    r = requests.get(server + ext)\n",
    "    \n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "\n",
    "#     decoded = r.content.decode()\n",
    "    dict_data = xmltodict.parse(r.content)\n",
    "    return dict_data['response']['result']['doc']['str'][1]['#text']\n",
    "    \n",
    "def human_from_entrez_to_ens(returned_entrez_id):\n",
    "    server = 'https://www.ncbi.nlm.nih.gov/'\n",
    "    ext = 'gene/' + str(returned_entrez_id)\n",
    "\n",
    "#     parser = etree.HTMLParser(recover=True)\n",
    "    page = requests.get(server + ext)\n",
    "    \n",
    "    if not page.ok:\n",
    "        page.raise_for_status()\n",
    "    dict_data = xmltodict.parse(page.content)\n",
    "    package = dict_data['html']['body']['div']['div']['form']['div'][0]['div'][3]['div']['div'][5]\n",
    "    returned_ensid = package['div'][1]['div'][0]['div']['div']['dl']['dd'][3]['a'][0]['#text']\n",
    "    \n",
    "    return returned_ensid.split(':')[-1]\n",
    "\n",
    "def it_child2(node, found):\n",
    "    if node == None:\n",
    "        return \n",
    "    if len(found) == 0:\n",
    "        for element in node.iterchildren():\n",
    "            if 'data-section' in element.attrib and element.attrib['data-section'] == 'Featured':\n",
    "                found.append(element.attrib['data-item-id'])\n",
    "            it_child2(element, found)\n",
    "    else:\n",
    "        return found\n",
    "    return found\n",
    "        \n",
    "def human_from_symbol_2entrez(human_symbol):\n",
    "    server = 'https://www.ncbi.nlm.nih.gov/'\n",
    "    ext = 'search/all/?term=' + 'TDH'\n",
    "    parser = etree.HTMLParser(recover=True)\n",
    "    page = requests.get(server + ext, headers={\"Content-Type\": \"application/json\"})\n",
    "\n",
    "    if not page.ok:\n",
    "        page.raise_for_status()\n",
    "\n",
    "    html = page.content.decode(\"utf-8\")\n",
    "    tree = etree.parse(StringIO(html), parser=parser)\n",
    "    root = tree.getroot()\n",
    "     \n",
    "    returned_entrez_id = None\n",
    "    returned_ens_id = None\n",
    "    \n",
    "    try:\n",
    "        found = []\n",
    "        returned_entrez_id = it_child2(root,found)[0].split(':')[-1]\n",
    "    except Exception:\n",
    "        pass\n",
    "    \n",
    "    try:\n",
    "        returned_ens_id = human_from_entrez_to_ens(returned_entrez_id)\n",
    "    except Exception:\n",
    "        pass\n",
    "    \n",
    "    return returned_entrez_id, returned_ens_id "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "species            object\n",
       "gene_id            object\n",
       "ensembl_symbol     object\n",
       "entrez_id          object\n",
       "ncbi_symbol        object\n",
       "vgnc_id            object\n",
       "vgnc_symbol        object\n",
       "hgnc_orthologs     object\n",
       "human_gene_id      object\n",
       "human_entrez_id    object\n",
       "hgnc_symbol        object\n",
       "dtype: object"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = pd.read_csv('/Users/liulihe95/EnrichKitDB/data/tmp/id_mapper/id_mapper_sus.txt')\n",
    "test2 = test.iloc[32200:32215,]\n",
    "test2 = test2.astype(str)\n",
    "test2.dtypes\n",
    "# test2.dtypes()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ensembl_symbol</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>ncbi_symbol</th>\n",
       "      <th>vgnc_id</th>\n",
       "      <th>vgnc_symbol</th>\n",
       "      <th>hgnc_orthologs</th>\n",
       "      <th>human_gene_id</th>\n",
       "      <th>human_entrez_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32200</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037847</td>\n",
       "      <td>MOB2</td>\n",
       "      <td>110259212.0</td>\n",
       "      <td>MOB2</td>\n",
       "      <td>VGNC:90291</td>\n",
       "      <td>MOB2</td>\n",
       "      <td>HGNC:24904</td>\n",
       "      <td>ENSG00000182208</td>\n",
       "      <td>81532.0</td>\n",
       "      <td>MOB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32201</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037848</td>\n",
       "      <td>nan</td>\n",
       "      <td>110255232.0</td>\n",
       "      <td>LOC110255232</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32202</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037854</td>\n",
       "      <td>L3HYPDH</td>\n",
       "      <td>110262303.0</td>\n",
       "      <td>L3HYPDH</td>\n",
       "      <td>VGNC:89612</td>\n",
       "      <td>L3HYPDH</td>\n",
       "      <td>HGNC:20488</td>\n",
       "      <td>ENSG00000126790</td>\n",
       "      <td>112849.0</td>\n",
       "      <td>L3HYPDH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32203</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037856</td>\n",
       "      <td>EDN2</td>\n",
       "      <td>102157561.0</td>\n",
       "      <td>EDN2</td>\n",
       "      <td>VGNC:87548</td>\n",
       "      <td>EDN2</td>\n",
       "      <td>HGNC:3177</td>\n",
       "      <td>ENSG00000127129</td>\n",
       "      <td>1907.0</td>\n",
       "      <td>EDN2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32204</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037858</td>\n",
       "      <td>nan</td>\n",
       "      <td>100518500.0</td>\n",
       "      <td>LOC100518500</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32205</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037865</td>\n",
       "      <td>SOX18</td>\n",
       "      <td>100049667.0</td>\n",
       "      <td>SOX18</td>\n",
       "      <td>VGNC:95510</td>\n",
       "      <td>SOX18</td>\n",
       "      <td>HGNC:11194</td>\n",
       "      <td>ENSG00000203883</td>\n",
       "      <td>54345.0</td>\n",
       "      <td>SOX18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32206</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037866</td>\n",
       "      <td>MAGEA10</td>\n",
       "      <td>110257696.0</td>\n",
       "      <td>MAGEA10</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32207</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037868</td>\n",
       "      <td>NPBWR2</td>\n",
       "      <td>110257396.0</td>\n",
       "      <td>NPBWR2</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32208</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037869</td>\n",
       "      <td>NDUFA5</td>\n",
       "      <td>100525371.0</td>\n",
       "      <td>NDUFA5</td>\n",
       "      <td>VGNC:110309</td>\n",
       "      <td>NDUFA5</td>\n",
       "      <td>HGNC:7688</td>\n",
       "      <td>ENSG00000128609</td>\n",
       "      <td>4698.0</td>\n",
       "      <td>NDUFA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32209</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037873</td>\n",
       "      <td>nan</td>\n",
       "      <td>110258905.0</td>\n",
       "      <td>LOC110258905</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32210</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037875</td>\n",
       "      <td>ZNF688</td>\n",
       "      <td>100515603.0</td>\n",
       "      <td>ZNF688</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32211</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037877</td>\n",
       "      <td>nan</td>\n",
       "      <td>110257991.0</td>\n",
       "      <td>LOC110257991</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32212</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037878</td>\n",
       "      <td>AHSP</td>\n",
       "      <td>102164536.0</td>\n",
       "      <td>AHSP</td>\n",
       "      <td>VGNC:85195</td>\n",
       "      <td>AHSP</td>\n",
       "      <td>HGNC:18075</td>\n",
       "      <td>ENSG00000169877</td>\n",
       "      <td>51327.0</td>\n",
       "      <td>AHSP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32213</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037880</td>\n",
       "      <td>NSMF</td>\n",
       "      <td>100514943.0</td>\n",
       "      <td>NSMF</td>\n",
       "      <td>VGNC:90912</td>\n",
       "      <td>NSMF</td>\n",
       "      <td>HGNC:29843</td>\n",
       "      <td>ENSG00000165802</td>\n",
       "      <td>26012.0</td>\n",
       "      <td>NSMF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32214</th>\n",
       "      <td>5</td>\n",
       "      <td>ENSSSCG00000037883</td>\n",
       "      <td>ssc-mir-9818</td>\n",
       "      <td>104796094.0</td>\n",
       "      <td>MIR9818</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      species             gene_id ensembl_symbol    entrez_id   ncbi_symbol  \\\n",
       "32200       5  ENSSSCG00000037847           MOB2  110259212.0          MOB2   \n",
       "32201       5  ENSSSCG00000037848            nan  110255232.0  LOC110255232   \n",
       "32202       5  ENSSSCG00000037854        L3HYPDH  110262303.0       L3HYPDH   \n",
       "32203       5  ENSSSCG00000037856           EDN2  102157561.0          EDN2   \n",
       "32204       5  ENSSSCG00000037858            nan  100518500.0  LOC100518500   \n",
       "32205       5  ENSSSCG00000037865          SOX18  100049667.0         SOX18   \n",
       "32206       5  ENSSSCG00000037866        MAGEA10  110257696.0       MAGEA10   \n",
       "32207       5  ENSSSCG00000037868         NPBWR2  110257396.0        NPBWR2   \n",
       "32208       5  ENSSSCG00000037869         NDUFA5  100525371.0        NDUFA5   \n",
       "32209       5  ENSSSCG00000037873            nan  110258905.0  LOC110258905   \n",
       "32210       5  ENSSSCG00000037875         ZNF688  100515603.0        ZNF688   \n",
       "32211       5  ENSSSCG00000037877            nan  110257991.0  LOC110257991   \n",
       "32212       5  ENSSSCG00000037878           AHSP  102164536.0          AHSP   \n",
       "32213       5  ENSSSCG00000037880           NSMF  100514943.0          NSMF   \n",
       "32214       5  ENSSSCG00000037883   ssc-mir-9818  104796094.0       MIR9818   \n",
       "\n",
       "           vgnc_id vgnc_symbol hgnc_orthologs    human_gene_id  \\\n",
       "32200   VGNC:90291        MOB2     HGNC:24904  ENSG00000182208   \n",
       "32201          nan         nan            nan              nan   \n",
       "32202   VGNC:89612     L3HYPDH     HGNC:20488  ENSG00000126790   \n",
       "32203   VGNC:87548        EDN2      HGNC:3177  ENSG00000127129   \n",
       "32204          nan         nan            nan              nan   \n",
       "32205   VGNC:95510       SOX18     HGNC:11194  ENSG00000203883   \n",
       "32206          nan         nan            nan              nan   \n",
       "32207          nan         nan            nan              nan   \n",
       "32208  VGNC:110309      NDUFA5      HGNC:7688  ENSG00000128609   \n",
       "32209          nan         nan            nan              nan   \n",
       "32210          nan         nan            nan              nan   \n",
       "32211          nan         nan            nan              nan   \n",
       "32212   VGNC:85195        AHSP     HGNC:18075  ENSG00000169877   \n",
       "32213   VGNC:90912        NSMF     HGNC:29843  ENSG00000165802   \n",
       "32214          nan         nan            nan              nan   \n",
       "\n",
       "      human_entrez_id hgnc_symbol  \n",
       "32200         81532.0        MOB2  \n",
       "32201             nan         nan  \n",
       "32202        112849.0     L3HYPDH  \n",
       "32203          1907.0        EDN2  \n",
       "32204             nan         nan  \n",
       "32205         54345.0       SOX18  \n",
       "32206             nan         nan  \n",
       "32207             nan         nan  \n",
       "32208          4698.0      NDUFA5  \n",
       "32209             nan         nan  \n",
       "32210             nan         nan  \n",
       "32211             nan         nan  \n",
       "32212         51327.0        AHSP  \n",
       "32213         26012.0        NSMF  \n",
       "32214             nan         nan  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_mapper_final = test2\n",
    "id_mapper_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "checking ensembl symbol for ENSSSCG00000037848\n",
      "checking ncbi human orthlogus for  110255232\n",
      "checking ensembl symbol for ENSSSCG00000037858\n",
      "checking ncbi human orthlogus for  100518500\n",
      "checking ncbi human orthlogus for  110257696\n",
      "parsing ncbi human orthlogus for 4109\n",
      "checking ncbi human orthlogus for  110257396\n",
      "checking ensembl symbol for ENSSSCG00000037873\n",
      "checking ncbi human orthlogus for  110258905\n",
      "checking ncbi human orthlogus for  100515603\n",
      "checking ensembl symbol for ENSSSCG00000037877\n",
      "checking ncbi human orthlogus for  110257991\n",
      "checking ncbi human orthlogus for  104796094\n"
     ]
    }
   ],
   "source": [
    "out = []\n",
    "for index, row in id_mapper_final.iterrows():\n",
    "# row = id_mapper_final.iloc[9]\n",
    "    if row['gene_id'] != 'nan':\n",
    "        if row['ensembl_symbol'] == 'nan':\n",
    "            # check ensembl_symbol\n",
    "            try:\n",
    "                row['ensembl_symbol'] = check_ensembl_symbol_helper(row['gene_id'])\n",
    "\n",
    "            except Exception:\n",
    "                pass\n",
    "\n",
    "        # check entrez id\n",
    "        if row['entrez_id'] == 'nan':\n",
    "            returned_entrez_id = None\n",
    "            returned_entrez_symbol = None\n",
    "\n",
    "            # ens 2 entrez: trust ncbi\n",
    "            try:\n",
    "                returned_entrez_id, returned_entrez_symbol = from_ens_2_entrez(row['gene_id'])\n",
    "\n",
    "            except Exception:\n",
    "                pass\n",
    "\n",
    "            if returned_entrez_id == None or returned_entrez_symbol == None:\n",
    "\n",
    "                # ens 2 entrez: trust ensembl\n",
    "                try:\n",
    "                    returned_entrez_id = from_ens_2_entrez_use_enblorg(row['gene_id'] )\n",
    "                    _, returned_entrez_symbol = from_ens_2_entrez(returned_entrez_id)\n",
    "\n",
    "                except Exception:\n",
    "                    pass\n",
    "\n",
    "            if not returned_entrez_id is None and not returned_entrez_symbol is None:\n",
    "                row['entrez_id'] = returned_entrez_id\n",
    "                row['ncbi_symbol'] = returned_entrez_symbol\n",
    "\n",
    "    # check human orthologus\n",
    "    if row['entrez_id'] != 'nan' and row['human_gene_id'] == 'nan':\n",
    "\n",
    "        try:\n",
    "            human_entrez_id = get_human_orth_ncbi(str(int(float(row['entrez_id']))))\n",
    "\n",
    "            row['human_entrez_id'] = human_entrez_id\n",
    "\n",
    "            human_symbol, hgnc, ens_id = parse_human_url(str(human_entrez_id))\n",
    "\n",
    "            row['hgnc_orthologs'] = hgnc\n",
    "\n",
    "            row['human_gene_id'] = ens_id\n",
    "\n",
    "            row['hgnc_symbol'] = human_symbol\n",
    "\n",
    "        except Exception:\n",
    "            pass\n",
    "\n",
    "    # check hgnc id and hgnc symbol\n",
    "    if row['hgnc_orthologs'] != 'nan' and row['hgnc_symbol'] == 'nan':\n",
    "        try:\n",
    "            row['hgnc_symbol'] = from_hgncid_to_hgncsymb(row['hgnc_orthologs'])\n",
    "        except Exception:\n",
    "            pass\n",
    "\n",
    "    if row['hgnc_symbol'] != 'nan' and (row['human_gene_id'] == 'nan' or row['human_entrez_id'] == 'nan'):\n",
    "\n",
    "        returned_entrez_id_human, returned_ens_id_human = human_from_symbol_2entrez(row['hgnc_symbol'])\n",
    "\n",
    "        if not returned_entrez_id_human is None:\n",
    "            row['human_entrez_id'] = returned_entrez_id_human\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "        if not returned_ens_id_human is None:\n",
    "            row['human_gene_id'] = returned_ens_id_human\n",
    "        else:\n",
    "            pass\n",
    "# pd.DataFrame(row).T\n",
    "# out.append(row)\n",
    "\n",
    "# pd.DataFrame(out)\n",
    "\n",
    "# test = pd.DataFrame(row)\n",
    "# test.to_csv(\n",
    "#     os.path.join('test',''.join([str(index),'.txt'])\n",
    "#     ), index=False, encoding='utf-8')\n",
    "\n",
    "#     id_mapper_final.to_csv(''.join(['new_', data_file_path]), index=False, encoding='utf-8')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "328500"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "2190 * 150"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ensembl_symbol</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>ncbi_symbol</th>\n",
       "      <th>vgnc_id</th>\n",
       "      <th>vgnc_symbol</th>\n",
       "      <th>hgnc_orthologs</th>\n",
       "      <th>human_gene_id</th>\n",
       "      <th>human_entrez_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32200</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004356</td>\n",
       "      <td>TRAPPC8</td>\n",
       "      <td>421095.0</td>\n",
       "      <td>TRAPPC8</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32201</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004358</td>\n",
       "      <td>GC</td>\n",
       "      <td>395696.0</td>\n",
       "      <td>GC</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32202</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004359</td>\n",
       "      <td>APOV1</td>\n",
       "      <td>396476.0</td>\n",
       "      <td>APOV1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32203</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004361</td>\n",
       "      <td>NPFFR2</td>\n",
       "      <td>428759.0</td>\n",
       "      <td>NPFFR2</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32204</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004362</td>\n",
       "      <td>FBXL4</td>\n",
       "      <td>421798.0</td>\n",
       "      <td>FBXL4</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32205</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004364</td>\n",
       "      <td>TSHR</td>\n",
       "      <td>428900.0</td>\n",
       "      <td>TSHR</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32206</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004365</td>\n",
       "      <td>TPD52</td>\n",
       "      <td>770339.0</td>\n",
       "      <td>TPD52</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32207</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004366</td>\n",
       "      <td>UNC5C</td>\n",
       "      <td>395101.0</td>\n",
       "      <td>UNC5C</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32208</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004367</td>\n",
       "      <td>TSPAN13</td>\n",
       "      <td>420595.0</td>\n",
       "      <td>TSPAN13</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32209</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004368</td>\n",
       "      <td>GJD4</td>\n",
       "      <td>420465.0</td>\n",
       "      <td>GJD4</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32210</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004370</td>\n",
       "      <td>ATP6V1H</td>\n",
       "      <td>426199.0</td>\n",
       "      <td>ATP6V1H</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32211</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004372</td>\n",
       "      <td>SYNJ2</td>\n",
       "      <td>421656.0</td>\n",
       "      <td>SYNJ2</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32212</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004373</td>\n",
       "      <td>MMS22L</td>\n",
       "      <td>421799.0</td>\n",
       "      <td>MMS22L</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32213</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004374</td>\n",
       "      <td>nan</td>\n",
       "      <td>418358.0</td>\n",
       "      <td>ILDR1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32214</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSGALG00010004377</td>\n",
       "      <td>AGR2</td>\n",
       "      <td>420596.0</td>\n",
       "      <td>AGR2</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      species             gene_id ensembl_symbol entrez_id ncbi_symbol  \\\n",
       "32200       3  ENSGALG00010004356        TRAPPC8  421095.0     TRAPPC8   \n",
       "32201       3  ENSGALG00010004358             GC  395696.0          GC   \n",
       "32202       3  ENSGALG00010004359          APOV1  396476.0       APOV1   \n",
       "32203       3  ENSGALG00010004361         NPFFR2  428759.0      NPFFR2   \n",
       "32204       3  ENSGALG00010004362          FBXL4  421798.0       FBXL4   \n",
       "32205       3  ENSGALG00010004364           TSHR  428900.0        TSHR   \n",
       "32206       3  ENSGALG00010004365          TPD52  770339.0       TPD52   \n",
       "32207       3  ENSGALG00010004366          UNC5C  395101.0       UNC5C   \n",
       "32208       3  ENSGALG00010004367        TSPAN13  420595.0     TSPAN13   \n",
       "32209       3  ENSGALG00010004368           GJD4  420465.0        GJD4   \n",
       "32210       3  ENSGALG00010004370        ATP6V1H  426199.0     ATP6V1H   \n",
       "32211       3  ENSGALG00010004372          SYNJ2  421656.0       SYNJ2   \n",
       "32212       3  ENSGALG00010004373         MMS22L  421799.0      MMS22L   \n",
       "32213       3  ENSGALG00010004374            nan  418358.0       ILDR1   \n",
       "32214       3  ENSGALG00010004377           AGR2  420596.0        AGR2   \n",
       "\n",
       "      vgnc_id vgnc_symbol hgnc_orthologs human_gene_id human_entrez_id  \\\n",
       "32200     nan         nan            nan           nan             nan   \n",
       "32201     nan         nan            nan           nan             nan   \n",
       "32202     nan         nan            nan           nan             nan   \n",
       "32203     nan         nan            nan           nan             nan   \n",
       "32204     nan         nan            nan           nan             nan   \n",
       "32205     nan         nan            nan           nan             nan   \n",
       "32206     nan         nan            nan           nan             nan   \n",
       "32207     nan         nan            nan           nan             nan   \n",
       "32208     nan         nan            nan           nan             nan   \n",
       "32209     nan         nan            nan           nan             nan   \n",
       "32210     nan         nan            nan           nan             nan   \n",
       "32211     nan         nan            nan           nan             nan   \n",
       "32212     nan         nan            nan           nan             nan   \n",
       "32213     nan         nan            nan           nan             nan   \n",
       "32214     nan         nan            nan           nan             nan   \n",
       "\n",
       "      hgnc_symbol  \n",
       "32200         nan  \n",
       "32201         nan  \n",
       "32202         nan  \n",
       "32203         nan  \n",
       "32204         nan  \n",
       "32205         nan  \n",
       "32206         nan  \n",
       "32207         nan  \n",
       "32208         nan  \n",
       "32209         nan  \n",
       "32210         nan  \n",
       "32211         nan  \n",
       "32212         nan  \n",
       "32213         nan  \n",
       "32214         nan  "
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_mapper_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# file_list = os.listdir('data/tmp/id_mapper/')\n",
    "# out = []\n",
    "# for item in file_list:\n",
    "#     if item.startswith('id_mapper'):\n",
    "#         print(item)\n",
    "#         tmp_df = pd.read_csv(os.path.join('data/tmp/id_mapper/',item))\n",
    "#         out.append(tmp_df)\n",
    "\n",
    "# all_id_mapper = pd.concat(out)\n",
    "# all_id_mapper = all_id_mapper.astype(str)\n",
    "# all_id_mapper_filter = all_id_mapper[all_id_mapper['gene_id'] != 'nan']\n",
    "# all_id_mapper_filter.reset_index(drop=True, inplace=True)\n",
    "# all_id_mapper_filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# file_list = os.listdir('data/tmp/id_mapper/')\n",
    "# out = []\n",
    "# for item in file_list:\n",
    "#     if item.startswith('id_mapper'):\n",
    "#         print(item)\n",
    "#         tmp_df = pd.read_csv(os.path.join('data/tmp/id_mapper/',item))\n",
    "#         out.append(tmp_df)\n",
    "\n",
    "# all_id_mapper = pd.concat(out)\n",
    "# all_id_mapper = all_id_mapper.astype(str)\n",
    "# all_id_mapper_filter = all_id_mapper[all_id_mapper['gene_id'] != 'nan']\n",
    "# all_id_mapper_filter.reset_index(drop=True, inplace=True)\n",
    "# all_id_mapper_filter\n",
    "\n",
    "# # count = 0\n",
    "# for index, row in all_id_mapper_filter.iterrows():\n",
    "#     if index % 2000 == 0:\n",
    "#         print('mark - ', index)\n",
    "# #     print(index)\n",
    "# #     count += 1\n",
    "# #     if count < 10:\n",
    "#     tmp_row_df = row.to_frame().T\n",
    "#     tmp_row_df.to_csv(os.path.join('id_raw',''.join([str(index),'.txt'])), index=False, encoding='utf-8')\n",
    "# #     else:\n",
    "# #         break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "checking ens symbol for  ENSOARG00000000038\n",
      "checking ens symbol for  ENSOARG00000000039\n",
      "checking ncbi entrez id and symbol for  ENSOARG00000000039\n",
      "checking ens symbol for  ENSOARG00000000041\n",
      "checking ncbi entrez id and symbol for  ENSOARG00000000041\n",
      "checking ens symbol for  ENSOARG00000000044\n",
      "checking ncbi entrez id and symbol for  ENSOARG00000000044\n",
      "checking ens symbol for  ENSOARG00000000045\n",
      "checking ncbi entrez id and symbol for  ENSOARG00000000045\n",
      "checking ens symbol for  ENSOARG00000000046\n",
      "checking ens symbol for  ENSOARG00000000048\n",
      "checking ens symbol for  ENSOARG00000000049\n",
      "checking ncbi entrez id and symbol for  ENSOARG00000000049\n",
      "checking ens symbol for  ENSOARG00000000050\n",
      "checking ncbi entrez id and symbol for  ENSOARG00000000050\n",
      "checking ens symbol for  ENSOARG00000000051\n",
      "checking ncbi entrez id and symbol for  ENSOARG00000000051\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ensembl_symbol</th>\n",
       "      <th>entrez_id</th>\n",
       "      <th>ncbi_symbol</th>\n",
       "      <th>vgnc_id</th>\n",
       "      <th>vgnc_symbol</th>\n",
       "      <th>hgnc_orthologs</th>\n",
       "      <th>human_gene_id</th>\n",
       "      <th>human_entrez_id</th>\n",
       "      <th>hgnc_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSOARG00000000038</td>\n",
       "      <td>FIBP</td>\n",
       "      <td>101110717</td>\n",
       "      <td>FIBP</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>HGNC:HGNC:3705</td>\n",
       "      <td>ENSG00000172500</td>\n",
       "      <td>9158</td>\n",
       "      <td>FIBP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSOARG00000000039</td>\n",
       "      <td>LOC101104933</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSOARG00000000041</td>\n",
       "      <td>None</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSOARG00000000044</td>\n",
       "      <td>IL2</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSOARG00000000045</td>\n",
       "      <td>LOC101105186</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSOARG00000000046</td>\n",
       "      <td>LHX1</td>\n",
       "      <td>101109722</td>\n",
       "      <td>LHX1</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>HGNC:HGNC:6593</td>\n",
       "      <td>ENSG00000273706</td>\n",
       "      <td>3975</td>\n",
       "      <td>LHX1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSOARG00000000048</td>\n",
       "      <td>PIGS</td>\n",
       "      <td>101120255</td>\n",
       "      <td>PIGS</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>HGNC:HGNC:14937</td>\n",
       "      <td>ENSG00000087111</td>\n",
       "      <td>94005</td>\n",
       "      <td>PIGS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSOARG00000000049</td>\n",
       "      <td>None</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSOARG00000000050</td>\n",
       "      <td>LOC101105686</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSOARG00000000051</td>\n",
       "      <td>None</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id ensembl_symbol  entrez_id ncbi_symbol vgnc_id  \\\n",
       "0  ENSOARG00000000038           FIBP  101110717        FIBP     nan   \n",
       "1  ENSOARG00000000039   LOC101104933        nan         nan     nan   \n",
       "2  ENSOARG00000000041           None        nan         nan     nan   \n",
       "3  ENSOARG00000000044            IL2        nan         nan     nan   \n",
       "4  ENSOARG00000000045   LOC101105186        nan         nan     nan   \n",
       "5  ENSOARG00000000046           LHX1  101109722        LHX1     nan   \n",
       "6  ENSOARG00000000048           PIGS  101120255        PIGS     nan   \n",
       "7  ENSOARG00000000049           None        nan         nan     nan   \n",
       "8  ENSOARG00000000050   LOC101105686        nan         nan     nan   \n",
       "9  ENSOARG00000000051           None        nan         nan     nan   \n",
       "\n",
       "  vgnc_symbol   hgnc_orthologs    human_gene_id human_entrez_id hgnc_symbol  \n",
       "0         nan   HGNC:HGNC:3705  ENSG00000172500            9158        FIBP  \n",
       "1         nan              nan              nan             nan         nan  \n",
       "2         nan              nan              nan             nan         nan  \n",
       "3         nan              nan              nan             nan         nan  \n",
       "4         nan              nan              nan             nan         nan  \n",
       "5         nan   HGNC:HGNC:6593  ENSG00000273706            3975        LHX1  \n",
       "6         nan  HGNC:HGNC:14937  ENSG00000087111           94005        PIGS  \n",
       "7         nan              nan              nan             nan         nan  \n",
       "8         nan              nan              nan             nan         nan  \n",
       "9         nan              nan              nan             nan         nan  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "id_mapper_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'end': 29502,\n",
       " 'canonical_transcript': 'ENSCHIT00000000044.1',\n",
       " 'biotype': 'protein_coding',\n",
       " 'seq_region_name': 'LWLT01000149.1',\n",
       " 'logic_name': 'ensembl',\n",
       " 'db_type': 'core',\n",
       " 'strand': 1,\n",
       " 'id': 'ENSCHIG00000000043',\n",
       " 'version': 1,\n",
       " 'assembly_name': 'ARS1',\n",
       " 'object_type': 'Gene',\n",
       " 'species': 'capra_hircus',\n",
       " 'source': 'ensembl',\n",
       " 'start': 20721}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "server = \"https://rest.ensembl.org\"\n",
    "ext = \"/lookup/id/ENSCHIG00000000043?\"\n",
    " \n",
    "r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    " \n",
    "if not r.ok:\n",
    "  r.raise_for_status()\n",
    "  sys.exit()\n",
    " \n",
    "decoded = r.json()\n",
    "decoded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_ens_symbol(ens_id):\n",
    "    server = \"https://rest.ensembl.org\"\n",
    "    ext = \"/lookup/id/\" + ens_id + \"?\"\n",
    "    r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    " \n",
    "    if r.ok:\n",
    "        decoded = r.json()\n",
    "#         r.raise_for_status()\n",
    "#         sys.exit()\n",
    "    if 'display_name' in decoded:\n",
    "        return decoded['display_name']\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = check_ens_symbol('ENSOARG00000008127')\n",
    "t is None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-19-94872ea8bea4>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-19-94872ea8bea4>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    https://www.ncbi.nlm.nih.gov/gene?cmd=Retrieve&list_uids=102176510\u001b[0m\n\u001b[0m           ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "https://www.ncbi.nlm.nih.gov/gene?cmd=Retrieve&list_uids=101110717"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'canonical_transcript': 'ENSCHIT00000000525.1',\n",
       " 'source': 'ensembl',\n",
       " 'object_type': 'Gene',\n",
       " 'display_name': 'SH2D4B',\n",
       " 'logic_name': 'ensembl',\n",
       " 'version': 1,\n",
       " 'description': 'SH2 domain containing 4B [Source:HGNC Symbol;Acc:HGNC:31440]',\n",
       " 'seq_region_name': '28',\n",
       " 'start': 9842112,\n",
       " 'end': 9922991,\n",
       " 'assembly_name': 'ARS1',\n",
       " 'strand': -1,\n",
       " 'id': 'ENSCHIG00000000351',\n",
       " 'biotype': 'protein_coding',\n",
       " 'species': 'capra_hircus',\n",
       " 'db_type': 'core'}"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# bta - 'display_name'\n",
    "# cap - 'display_name'\n",
    "# sus - 'display_name'\n",
    "# ovi - 'display_name'\n",
    "\n",
    "\n",
    "\n",
    "# decoded\n",
    "# for item in decoded:\n",
    "#     print(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "entrez_id: 101105454 gene symbol:  SIRT3\n"
     ]
    }
   ],
   "source": [
    "# r = requests.get('https://www.ncbi.nlm.nih.gov/gene/?term=ENSOARG00000019090', headers={ \"Content-Type\" : \"application/json\"})\n",
    "target_url = 'https://www.ncbi.nlm.nih.gov/gene/?term=' + 'ENSOARG00000008832'\n",
    "response = requests.get(target_url)\n",
    "dict_data = xmltodict.parse(response.content)\n",
    "package = dict_data['html']['body']['div']['div']['form']['div'][0]['div'][3]['div']['div'][5]\n",
    "entrez_id = package['div'][0]['div'][0]['span']['#text'].split(' ')[2][:-1]\n",
    "name = package['div'][1]['div'][0]['div']['div']['dl']['dd'][0]['#text']\n",
    "print('entrez_id:',entrez_id.split(' ')[0], 'gene symbol: ', name.split(' ')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'http://uswest.ensembl.org/Ovis_aries/Gene/Summary?g=ENSOARG00000000038;r=21:43305834-43310013;t=ENSOART00000000039'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Element html at 0x7fe6e087a5f0>"
      ]
     },
     "execution_count": 294,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# target_url = 'https://uswest.ensembl.org/Multi/Search/Results?q=ENSOARG00000000038;site=ensembl'\n",
    "target_url = 'https://uswest.ensembl.org/Gene/Summary?g=ENSOARG00000000041'\n",
    "# target_url = 'http://uswest.ensembl.org/Ovis_aries/Gene/Summary?g=ENSOARG00000000038;r=21:43305834-43310013;t=ENSOART00000000039'\n",
    "# response = requests.get(target_url)\n",
    "# dir(response)\n",
    "# for line in response.iter_lines():\n",
    "#     print(line)\n",
    "# t = response.content.decode()\n",
    "# t\n",
    "# dict_data = xmltodict.parse(response.text)\n",
    "# response.text.split('\\n')\n",
    "\n",
    "page = requests.get(target_url)\n",
    "# page.content.decode()\n",
    "# parser = etree.HTMLParser(recover=True)\n",
    "\n",
    "# html = etree.HTML(page.content, parser)\n",
    "\n",
    "\n",
    "# Decode the page content from bytes to string\n",
    "html = page.content.decode(\"utf-8\")\n",
    "\n",
    "# Create your etree with a StringIO object which functions similarly\n",
    "# to a fileHandler\n",
    "tree = etree.parse(StringIO(html), parser=parser)\n",
    "root = tree.getroot()\n",
    "root\n",
    "# # Call this function and pass in your tree\n",
    "# def get_links(tree):\n",
    "#     # This will get the anchor tags <a href...>\n",
    "#     refs = tree.xpath(\"//a\")\n",
    "#     # Get the url from the ref\n",
    "#     links = [link.get('href', '') for link in refs]\n",
    "#     # Return a list that only ends with .com.br\n",
    "#     return [l for l in links if l.endswith('.com.br')]\n",
    "\n",
    "\n",
    "# # Example call\n",
    "# links = get_links(tree)\n",
    "def it_child(node):\n",
    "    for element in node.iterchildren():\n",
    "        if element.tag == 'a' and 'href' in element.attrib and element.attrib['href'].startswith('http://www.ncbi.nlm.nih.gov/entrez'): \n",
    "            print(element.text)\n",
    "        it_child(element)\n",
    "        \n",
    "it_child(root)\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
